### PR TITLE
Report task input hashes when caching

### PR DESCRIPTION
**Pull request by wolfs**
Add a listener which can be used to capture input hashes created by
task caching. We use that listener to log the hashes on the info log
level.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/TaskArtifactState.java [hunk 1]
```diff
@@ -41,10 +41,10 @@
     boolean isAllowedToUseCachedResults();
 
     /**
-     * Returns the calculated cache key for the task's current state, or {@code null} if the task is not cacheable.
+     * Initializes and returns the current execution.
      */
     @Nullable
-    BuildCacheKey calculateCacheKey();
+    TaskExecution getCurrentExecution();
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/TaskArtifactState.java]: Can we keep `TaskExecution` hidden and just let this be a pass-through to get the `BuildCacheKey`?

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesListener.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import org.gradle.api.Task;
+import org.gradle.caching.BuildCacheKey;
+
+import java.util.Map;
+import java.util.Set;
+
+public interface TaskCachingHashesListener {
+    void inputsCollected(Task task, BuildCacheKey key, TaskCachingInputs hashes);
+
+    class TaskCachingInputs {
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesListener.java]: Since we're already wrapping things into a richer type, could we just return `String` forms of the hashes? Do they need to be the byte forms?

Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesListener.java]: Maybe `BuildCacheKeyInputs`?

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesListener.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import org.gradle.api.Task;
+import org.gradle.caching.BuildCacheKey;
+
+import java.util.Map;
+import java.util.Set;
+
+public interface TaskCachingHashesListener {
+    void inputsCollected(Task task, BuildCacheKey key, TaskCachingInputs hashes);
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesListener.java]: nit, I like the order:

`inputsCollected(BuildCacheKey key, TaskCachingInputs hashes, Task task)`

Because it flows a bit like, "inputs collected for `key` are `hashes` from `task`"

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesLogger.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+import java.util.Map;
+
+public class TaskCachingHashesLogger implements TaskCachingHashesListener {
+    private final Logger logger = LoggerFactory.getLogger(SkipCachedTaskExecuter.class);
+
+    @Override
+    public void inputsCollected(Task task, BuildCacheKey key, TaskCachingInputs hashes) {
+        logger.info("Cache key for {} is {}", task, key);
+        logger.info("{} classloader hash: {}", task, hashToStringRepresentation(hashes.getClassLoaderHash()));
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesLogger.java]: Maybe we should make all of these logs below here debug level?

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/VerifyNoInputChangesTaskExecuter.java [hunk 5]
```diff
@@ -29,18 +30,20 @@
...
     }
 
     @Override
     public void execute(TaskInternal task, TaskStateInternal state, TaskExecutionContext context) {
-        BuildCacheKey beforeExecution = context.getTaskArtifactState().calculateCacheKey();
+        BuildCacheKey beforeExecution = taskCacheKeyCalculator.calculate(context.getTaskArtifactState().getCurrentExecution(), task);
         delegate.execute(task, state, context);
         if (beforeExecution != null) {
-            BuildCacheKey afterExecution = repository.getStateFor(task).calculateCacheKey();
+            BuildCacheKey afterExecution = taskCacheKeyCalculator.calculate(repository.getStateFor(task).getCurrentExecution(), task);
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/VerifyNoInputChangesTaskExecuter.java]: I think this means we'd notify listeners twice when this is enabled?

Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/VerifyNoInputChangesTaskExecuter.java]: nevermind.  I see below it doesn't have a listener attached.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipCachedTaskExecuter.java [hunk 6]
```diff
@@ -81,8 +89,7 @@ public void execute(final TaskInternal task, TaskStateInternal state, TaskExecut
                         cacheable = true;
                         TaskArtifactState taskState = context.getTaskArtifactState();
                         try {
-                            cacheKey = taskState.calculateCacheKey();
-                            LOGGER.info("Cache key for {} is {}", task, cacheKey);
+                            cacheKey = cacheKeyCalculator.calculate(taskState.getCurrentExecution(), task);
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipCachedTaskExecuter.java]: I think the cache key calculator should be wired in through the `TaskArtifactStateRepository` and in `TaskArtifactState`, we could save the result of calculating the `BuildCacheKey`.  That way there would be fewer services required here and `SkipCachedTaskExecutor` would be doing less.

This might make wiring in the listener different since you wouldn't know how the cache key was about to be used.  Maybe that would let us build a better listener that compares the before and after keys at a deeper level?

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/changes/NoHistoryArtifactState.java [hunk 7]
```diff
@@ -42,7 +42,7 @@ public boolean isAllowedToUseCachedResults() {
     }
 
     @Override
-    public BuildCacheKey calculateCacheKey() {
+    public TaskExecution getCurrentExecution() {
         return null;
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/changes/NoHistoryArtifactState.java]: (In combination with what I suggested above that we should be returning a `BuildCacheKey` here).  We should make it so we never return a null `BuildCacheKey`.  We should have a "no cache key" value that we use instead.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipCachedTaskExecuter.java (hunk 6 above)
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipCachedTaskExecuter.java]: We also need to make this so we're always creating the cache key for all tasks and all the time, not just the ones that have caching enabled.

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/TaskExecutionServices.java [hunk 8]
```diff
@@ -139,17 +143,20 @@ TaskExecuter createTaskExecuter(TaskArtifactStateRepository repository, TaskOutp
         );
     }
 
-    private static TaskExecuter createSkipCachedExecuterIfNecessary(StartParameter startParameter, BuildCacheConfigurationInternal buildCacheConfiguration, TaskOutputPacker packer, TaskOutputsGenerationListener taskOutputsGenerationListener, TaskOutputOriginFactory taskOutputOriginFactory, TaskExecuter delegate) {
+    private static TaskExecuter createSkipCachedExecuterIfNecessary(StartParameter startParameter, BuildCacheConfigurationInternal buildCacheConfiguration, TaskOutputPacker packer, TaskOutputsGenerationListener taskOutputsGenerationListener, TaskOutputOriginFactory taskOutputOriginFactory, ListenerManager listenerManager, TaskExecuter delegate) {
         if (startParameter.isTaskOutputCacheEnabled()) {
-            return new SkipCachedTaskExecuter(taskOutputOriginFactory, buildCacheConfiguration, packer, taskOutputsGenerationListener, delegate);
+            TaskCachingHashesListener hashesListener = listenerManager.getBroadcaster(TaskCachingHashesListener.class);
+            TaskCacheKeyCalculator taskCacheKeyCalculator = new TaskCacheKeyCalculator(hashesListener);
+            listenerManager.addListener(new TaskCachingHashesLogger());
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/internal/service/scopes/TaskExecutionServices.java]: This is another reason for trying to push the cache key calculator out of `SkipCachedTaskExecutor`.  There's getting to be too many things going into it.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+    public TaskCacheKeyCalculator() {
+        this(null);
+    }
+
+    public TaskCacheKeyCalculator(TaskCachingHashesListener hashesListener) {
+        this.hashesListener = hashesListener;
+    }
+
+    public BuildCacheKey calculate(TaskExecution execution, Task task) {
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java]: I think we should pull out the pieces we need out of `TaskExecution` vs passing a `TaskExecution` here

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+            addInputProperty(entry.getKey(), hash, builder, inputHashes);
+        }
+
+        Set<String> outputPropertyNames = new HashSet<String>();
+        for (String cacheableOutputPropertyName : sortStrings(execution.getOutputPropertyNamesForCacheKey())) {
+            outputPropertyNames.add(cacheableOutputPropertyName);
+            builder.putString(cacheableOutputPropertyName);
+        }
+
+        BuildCacheKey cacheKey = builder.build();
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java]: And we need to clean out of all of this code from `TaskExecution` too?

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+        this(null);
+    }
+
+    public TaskCacheKeyCalculator(TaskCachingHashesListener hashesListener) {
+        this.hashesListener = hashesListener;
+    }
+
+    public BuildCacheKey calculate(TaskExecution execution, Task task) {
+        if (execution == null || execution.getTaskActionsClassLoaderHash() == null || execution.getTaskActionsClassLoaderHash() == null) {
+            hashesListener.inputsCollected(task, null, new TaskCachingHashesListener.TaskCachingInputs(
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java]: as above, I think we should have a "no cache key" `BuildCacheKey` that we use in cases of tasks that cannot have a cache key calculated

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    private static List<String> sortStrings(Collection<String> entries) {
+        List<String> sortedEntries = Lists.newArrayList(entries);
+        Collections.sort(sortedEntries);
+        return sortedEntries;
+    }
+
+    private static HashCode hashForObject(Object object) {
+        return ((DefaultBuildCacheKeyBuilder) new DefaultBuildCacheKeyBuilder().appendToCacheKey(object)).buildHashCode();
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java]: Could we change `appendToCacheKey` in `DefaultBuildCacheKeyBuilder` to just be a static method that we can call (that returns a hash)?

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    public TaskCacheKeyCalculator() {
+        this(null);
+    }
+
+    public TaskCacheKeyCalculator(TaskCachingHashesListener hashesListener) {
+        this.hashesListener = hashesListener;
+    }
+
+    public BuildCacheKey calculate(TaskExecution execution, Task task) {
+        if (execution == null || execution.getTaskActionsClassLoaderHash() == null || execution.getTaskActionsClassLoaderHash() == null) {
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java]: Would it be useful to see all of the other inputs even if the classloaders are null?

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesListener.java (hunk 2 above)
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesListener.java]: My thinking would be that I want to use a little memory as possible. Generating the strings is not free cpu wise - my thinking was that we should just pass around the `byte[]`. But we can use the `String`s and then have a look at the performance tests.

Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesListener.java]: At some later point we probably also will publish the output hashes. These probably won't have to do anything with the `BuildCacheKey`. Still, we probably should come up with a better name.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesLogger.java (hunk 4 above)
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesLogger.java]: I am not very passionate about that. I chose info since we also are going to publish the hashes to cloud services - so info seemed adequate. We are already publishing all the single hashes in [DefaultBuildCacheKeyBuilder](https://github.com/gradle/gradle/blob/0eeac61ab99137123cf3c8bc2f6e0ab407c039ce/subprojects/core/src/main/java/org/gradle/caching/internal/DefaultBuildCacheKeyBuilder.java#L137-L137) on the debug level. This is another reason why I chose info level here. WDYT?

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java (hunk 9 above)
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java]: TaskExecution consists exactly of the pieces we need for calculating the build key. The only additional data is `declaredOutputFilePaths`. Which other pieces do you see you wouldn't want to pass around?

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java (hunk 13 above)
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java]: That is probably such a fundamental problem that the other inputs just become meaningless. WDYT?

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesListener.java (hunk 2 above)
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesListener.java]: Maybe just make the methods returns `String` then? I was thinking it was just ugly to be carrying the byte[] around.

Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesListener.java]: Maybe that's just another method for this listener?

`outputsProduced(BuildCacheKey, BuildCacheOutputs, Task)`

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesLogger.java (hunk 4 above)
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesLogger.java]: Hrm, I don't have a strong feeling either.  I just don't want to put more noise in `--info`.

Do we want this at all?  Maybe we should remove some logging that's already producing this same information (e.g., from `TaskTypeTaskStateChanges` or the cache key builder?)

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java (hunk 9 above)
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java]: I don't like that it's mutable, but I was thinking about `LazyTaskExecution` more than just `TaskExecution`.  It's all really hard to follow when a particular value is going to be populated or not.  

I don't feel very strongly about this.  I also didn't want `TaskExecution` being used too widely (again, since it's hard to tell when something might be added/modified).

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java (hunk 13 above)
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java]: I think you're right.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesListener.java (hunk 2 above)
Comment (adammurdoch) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesListener.java]: byte[] or String are not great options as either forces the hashes to be coerced from whatever internal representation happens to be most efficient for the hash, into something that is not.

We should be using strong types for hashes in all our APIs. These strong types should have query methods that return string or byte[] representation or whatever form the consumer might like the hash represented as. Currently, we're using guava's `HashCode` type. I would continue to do so here.

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/TaskExecutionServices.java [hunk 14]
```diff
@@ -139,17 +143,20 @@ TaskExecuter createTaskExecuter(TaskArtifactStateRepository repository, TaskOutp
         );
     }
 
-    private static TaskExecuter createSkipCachedExecuterIfNecessary(StartParameter startParameter, BuildCacheConfigurationInternal buildCacheConfiguration, TaskOutputPacker packer, TaskOutputsGenerationListener taskOutputsGenerationListener, TaskOutputOriginFactory taskOutputOriginFactory, TaskExecuter delegate) {
+    private static TaskExecuter createSkipCachedExecuterIfNecessary(StartParameter startParameter, BuildCacheConfigurationInternal buildCacheConfiguration, TaskOutputPacker packer, TaskOutputsGenerationListener taskOutputsGenerationListener, TaskOutputOriginFactory taskOutputOriginFactory, ListenerManager listenerManager, TaskExecuter delegate) {
         if (startParameter.isTaskOutputCacheEnabled()) {
```
Comment (adammurdoch) [inline subprojects/core/src/main/java/org/gradle/internal/service/scopes/TaskExecutionServices.java]: We should do this regardless of whether the cache is being used. This information is also very valuable when the cache is not being used.

We should also sync this up with the 'out-of-date' messages, so the same information is presented there.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesListener.java (hunk 3 above)
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesListener.java]: I looked at other listeners like `TaskExecutionListener` and `TaskInputsListener` and it seems that `Task` is always the first argument to the listener methods. I would also read the line `inputs collected for task have the cache key `key` and hashes `hashes`.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesListener.java (hunk 2 above)
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingHashesListener.java]: I changed it to use `HashCode`.

**Comment by big-guy** [comment]
@wolfs I think we want to say that this isn't going to make it for 3.4?

**Comment by big-guy** [comment]
@wolfs Let's merge this and get the Crucible review going.  I don't see anything that stops us from getting this into master now.
